api/rest/client/client.go

```diff
@@ -366,3 +386,45 @@ func isUnixSocketAddress(addr ma.Multiaddr) bool {
 	value, err := addr.ValueForProtocol(ma.P_UNIX)
 	return (value != "" && err == nil)
 }
+
+// isDNSAddress returns if the given address corresponds to a
+// DNSADDR protocol type
+func isDNSAddress(addr ma.Multiaddr) bool {
+	if addr == nil {
+		return false
+	}
+	dnsaddr, err := addr.ValueForProtocol(ma.P_DNSADDR)
+	return (dnsaddr != "" && err == nil)
+}
+
+// resolve dnsaddr
+func resolveDNSAddr(addr ma.Multiaddr) ([]ma.Multiaddr, error) {
+	var prevResolved []ma.Multiaddr
+	var currResolved []ma.Multiaddr
+
+	// Only resolve dnsaddr
+	if !isDNSAddress(addr) {
+		// return the addr as the entry to be resolved later
+		currResolved = append(currResolved, addr)
+		return currResolved, nil
+	}
+
+	// recursively resolve
+	currResolved = append(currResolved, addr)
+	for len(currResolved) != len(prevResolved) {
+		var tobeResolved []ma.Multiaddr
+		for _, addr := range currResolved {
+			resolveCtx, cancel := context.WithTimeout(context.Background(), ResolveTimeout)
+			defer cancel()
+			resolved, err := madns.Resolve(resolveCtx, addr)
+			if err != nil {
+				return []ma.Multiaddr{}, err
+			}
+			tobeResolved = append(tobeResolved, resolved...)
+		}
+		prevResolved = currResolved
+		currResolved = tobeResolved
```

Comment:
Resolve once. This is a bit confusing. Even if recursive resolution would bring something we may just want to load balance on the first-resolved addresses only.

Reply:
oh alright. I was looking at cases like this https://github.com/multiformats/multiaddr/blob/master/protocols/DNSADDR.md where the root dnsaddr resolves into other dnsaddrs

Reply:
Let's take the simple approach now. dnsaddr addresses that resolve to other dnsaddr addresses is something I haven't seen used in reality, and if it was perhaps it would be `madns` problem to handle recursive resolution.